packages/synapse-react-client/src/components/EntityUpload/EntityUpload.tsx

```diff
@@ -0,0 +1,321 @@
+  }
+
+  const numberOfItemsCompleted = uploadProgress.filter(
+    p => p.status === 'COMPLETE',
+  ).length
+
+  function uploadFileList(fileList: ArrayLike<File>) {
+    const args = Array.from(fileList).map(file => {
+      if (isFileEntity) {
+        return { file, existingEntityId: entityId }
+      }
+      return {
+        file,
+        rootContainerId: entityId,
+      }
+    })
+    initiateUpload(args)
+  }
+
+  useImperativeHandle(ref, () => ({
+    handleUploads: uploadFileList,
+  }))
+
+  return (
+    <div>
+      <EntityUploadPromptDialog activePrompts={activePrompts} />
+      <ExternalObjectStoreCredentialsForm
+        uploadDestination={uploadDestination}
+        accessKey={accessKey}
+        setAccessKey={setAccessKey}
+        secretKey={secretKey}
+        setSecretKey={setSecretKey}
+      />
+      <Stack
+        sx={{
+          width: '100%',
+          height: '235px',
+          border: '1px dashed #D9D9D9',
+          backgroundColor: 'grey.100',
+          textAlign: 'center',
+        }}
+        justifyContent={'center'}
+        alignItems={'center'}
+      >
+        {(isPrecheckingUpload || isLoading) && (
+          <>
+            <div>
+              <SynapseSpinner size={40} />
+            </div>
+            <Typography variant={'smallText1'} my={2}>
+              {isLoading ? 'Loading...' : 'Preparing files for upload...'}
+            </Typography>
+          </>
+        )}
+        {!isPrecheckingUpload && !isLoading && (
+          <>
+            <img
+              src={
+                'https://s3.us-east-1.amazonaws.com/static.synapse.org/images/upload_illustration.svg'
+              }
+            />
+            {/* File input */}
+            <input
+              className="form-control"
+              type="file"
+              id="filesToUpload"
+              name="file upload"
+              style={{ display: 'none' }}
+              aria-hidden="true"
+              multiple={!isFileEntity}
+              ref={fileInputRef}
+              onChange={e => {
+                if (e.target.files != null) {
+                  uploadFileList(e.target.files)
+                }
+              }}
+            />
+            {/* Folder input */}
+            <input
+              className="form-control"
+              type="file"
+              id="folderToUpload"
+              name="folder upload"
+              style={{ display: 'none' }}
+              aria-hidden="true"
+              ref={folderInputRef}
+              onChange={e => {
+                if (e.target.files != null) {
+                  uploadFileList(e.target.files)
+                }
+              }}
+              // @ts-expect-error - webkitdirectory is not included in the React.InputHTMLAttributes type
+              webkitdirectory="true"
+            />
+
+            <Typography variant={'body1'} my={2}>
+              <Link onClick={handleClick}>Click to upload</Link> or drag and
+              drop
```

Comment:
Note: no drag-n-drop handler was added here; still relying on the SWC implementation. It must be re-implemented before it will work in other apps